### PR TITLE
new symbion features for advanced malware analysis

### PR DESCRIPTION
**Pull request by degrigis**
This PR implements the following things after I spent some time in the past few months analyzing real-world malware with angr + Symbion:

1. The possibility to change registers' values directly from the Symbion Exploration Technique interface.

2. Couple of Windows SimProcedures: GetLastInfoInput, GetProcessAffinityMask, gethostbyname, HeapFree.

3) A better way to synchronize the SimProcedures addresses with the concrete process in the concrete plugin.

4) The possibility to decide to synchronize only the functions for which we have SimProcedures, avoiding to hook stubs for the one we are missing. This feature has been requested on our slack by @julioauto and I thought was useful to have. ( the parameter `stubs_on_sync` of the Concrete plugin constructor is controlling this behavior ).

### CODE REVIEW

File: angr/procedures/win32/GetLastInputInfo.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+class GetLastInputInfo(angr.SimProcedure):
+    cbSize = None
+    dwTime = None
+
+    def run(self, plii):
+        self.fill_symbolic()
+        l.warn("Setting symbolic memory at %s", str(plii))
+
+        self.state.mem[plii].dword = self.cbSize
+        self.state.mem[plii].dword = self.dwTime
```
Comment (rhelmot) [inline angr/procedures/win32/GetLastInputInfo.py]: do youmean plii + 4?

File: angr/sim_manager.py [hunk 2]
```diff
@@ -314,7 +315,7 @@ def step(self, stash='active', n=None, selector_func=None, step_func=None,
         :returns:           The simulation manager, for chaining.
         :rtype:             SimulationManager
         """
-        l.info("Stepping %s of %s", stash, self)
+        #l.info("Stepping %s of %s", stash, self)
```
Comment (rhelmot) [inline angr/sim_manager.py]: uncomment plz

File: angr/sim_manager.py [hunk 3]
```diff
@@ -27,6 +27,7 @@ class SimulationManager:
 
     Note that you shouldn't usually be constructing SimulationManagers directly - there is a convenient shortcut for
     creating them in ``Project.factory``: see :class:`angr.factory.AngrObjectFactory`.
+    creating them in ``Project.factory``: see :class:`angr.factory.AngrObjectFactory`.
```
Comment (rhelmot) [inline angr/sim_manager.py]: dup?

File: angr/knowledge_plugins/labels.py [hunk 4]
```diff
@@ -11,14 +11,15 @@ def __init__(self, kb):
...
         for obj in kb._project.loader.all_objects:
-            for v in obj.symbols:
-                if v.name and not v.is_import and v.type not in {cle.SymbolType.TYPE_OTHER, }:
-                    self._labels[v.rebased_addr] = v.name
-                    self._reverse_labels[v.name] = v.rebased_addr
             try:
+                for v in obj.symbols:
+                    if v.name and not v.is_import and v.type not in {cle.Symbol.TYPE_OTHER, }:
+                        self._labels[v.rebased_addr] = v.name
+                        self._reverse_labels[v.name] = v.rebased_addr
```
Comment (rhelmot) [inline angr/knowledge_plugins/labels.py]: What part of this will explode?

Comment (degrigis) [reply angr/knowledge_plugins/labels.py]: weird, I moved the `try` back to where it was before and the issue I had disappeared...
well, if I will ever reproduce the issue I will get back to this.

File: angr/engines/concrete.py [hunk 5]
```diff
@@ -42,30 +42,33 @@ def _check(self, state, *args, **kwargs):
...
-        :param state:               The state with which to execute
+        :param state:               The state with which to execute.
         :param extra_stop_points:   list of a addresses where to stop the concrete execution and return to the
-                                    simulated one
-        :param concretize:          list of tuples (address, symbolic variable) that are going to be written
-                                    in the concrete process memory
+                                    simulated one.
+        :param memory_concretize:   list of tuples (address, symbolic variable) that are going to be written
+                                    in the concrete process memory.
+        :param register_concretize  list of tuples (reg_name, symbolic variable) that are going to be written
```
Comment (julioauto) [inline angr/engines/concrete.py]: Is the lack of a ':' after 'register_concretize' what is making the CI fail?
```suggestion
        :param register_concretize:  list of tuples (reg_name, symbolic variable) that are going to be written
```

Comment (degrigis) [reply angr/engines/concrete.py]: Seems no, I actually forgot that this was not already merged, I'll talk with @rhelmot! Thanks for the heads up!

**Pull request closed without merging**